src/hardware/mouse/mouse.cpp

```diff
@@ -37,14 +37,260 @@
 
 CHECK_NARROWING();
 
-bool seamless_driver  = false;
-bool seamless_setting = false;
-
 static Bitu int74_ret_callback = 0;
 
 static MouseQueue &mouse_queue  = MouseQueue::GetInstance();
 static ManyMouseGlue &manymouse = ManyMouseGlue::GetInstance();
 
+// ***************************************************************************
+// GFX-related decision making
+// ***************************************************************************
+
+static struct {
+	bool fullscreen   = false; // if full screen mode is active
+	uint16_t clip_x   = 0;     // clipping = size of black border (one side)
+	uint16_t clip_y   = 0;
+
+	bool captured     = false; // if GFX was requested to capture mouse
+	bool visible      = false; // if GFX was requested to make cursor visible
+	bool raw_input    = false; // if GFX was requested to provide raw movements
+
+	bool has_focus    = false; // if our window has focus
+	bool gui_takeover = false; // if a GUI requested to take over the mouse
+	bool user_capture = false; // if user requested mouse to be captured
+
+	bool seamless     = false; // if seamless mouse integration is in effect
+	bool drop_events  = true;  // if we should drop mouse events
+
+	bool capture_on_click  = false; // if any button click should capture the mouse
+	bool capture_on_middle = false; // if middle button press should capture the mouse
+	bool release_on_middle = false; // if middle button press should release the mouse
+	bool toggle_on_hotkey  = false; // if hotkey should toggle mouse capture
+
+	bool cursor_outside   = true; // if mouse cursor is outside of drawing area
+	uint16_t cursor_x_abs = 0;    // absolute position from start of drawing area
+	uint16_t cursor_y_abs = 0;
+
+} mouse_gfx;
+
+static void update_cursor_absolute_position(const int32_t x_abs, const int32_t y_abs)
+{
+	mouse_gfx.cursor_outside = false;
+
+	auto calculate = [&](const int32_t absolute,
+	                     const uint16_t clipping,
+	                     const uint16_t resolution) -> uint16_t {
+		assert(resolution > 1u);
+		assert(clipping * 2 < resolution);
+
+		if (absolute < clipping) {
+			mouse_gfx.cursor_outside = true;
+			return 0;
+		} else if (absolute >= resolution - clipping) {
+			mouse_gfx.cursor_outside = true;
+			return static_cast<uint16_t>(resolution - 1);
+		}
+
+		const auto result = absolute - clipping;
+		return static_cast<uint16_t>(result);
+	};
+
+	auto &x = mouse_gfx.cursor_x_abs;
+	auto &y = mouse_gfx.cursor_y_abs;
+
+	x = calculate(x_abs, mouse_gfx.clip_x, mouse_shared.resolution_x);
+	y = calculate(y_abs, mouse_gfx.clip_y, mouse_shared.resolution_y);
+}
```

Comment:
I understand we're trying to match SDL's use of 32-bit coordinates (by not casting down to 16-bit), which lets us handle future large absolutely resolutions or very high DPI screens. 

But I see here our `cursor_x_abs` and clipping coordinates getting truncated or wrapped back down to 16-bit. This cuts-short our ability to handle SDL's > 16-bit screen axis (which is fine; that's in the distant future!).

Packing coordinates to 16-bit is fine, but we should probably assert those limits before casting; suggest `check_cast<uint16_t>(resolution - 1)` and `check_cast<uint16_t>(result)`, so future maintainers can easily catch these limits in debug builds.

`check_cast`:
 - asserts the value won't wrap in debug builds
 - becomes a fast `static_assert` in release builds

Reply:
I did a quick research - it seems that resolutions over 16-bit capabilities are not that distant future (see [Wikipedia article](https://en.wikipedia.org/wiki/16K_resolution)). I've changed the code to accept resolutions above 65535 - the performance impact should be negligible nevertheless.

Reply:
:+1: ; nice investigation!